Adding DocTest to TrOCR

arnaudstiegler opened the pull request:
# Adding TrOCR to DocTests

For this model, there was actually a single docstring in the entire file (for the forward method).

A couple of comments:
- As far as I'm aware, there is no TF version of this model
- TrOCR is an edge case because it's meant to be used as the decoder for a VisionEncoderDecoder. So the forward function of the TrOCR is not meant to be called directly. As a result, I gave some example code to run a forward pass with TrOCR within a VisionEncoderDecoder 

Let me know if the docstring is relevant to the problem. I can also revert adapt it to actually just showcase the forward for the `TrOCRForCausalLM` outside of a VisionEncoderDecoder

@patrickvonplaten @ydshieh @patil-suraj

arnaudstiegler committed:
docstring still WIP | adding to documentation_tests

arnaudstiegler committed:
clean version | passes tests

arnaudstiegler committed:
adding to documentation_test

arnaudstiegler committed:
adding forward for training pass

arnaudstiegler committed:
make fixup applied

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_16398). All of your documentation changes will be reflected on that endpoint.

ydshieh commented:
I took a quick look for now, and think it is indeed a nice addition, considering there is currently no example in `modeling_trocr.py` at all.
Thank you, @arnaudstiegler!

ydshieh commented on src/transformers/models/trocr/modeling_trocr.py:
````diff
@@ -891,13 +891,40 @@ def forward(
         Example:
 
         ```python
+        >>> from transformers import TrOCRProcessor, VisionEncoderDecoderModel
````
Would be great to avoid import `VisionEncoderDecoderModel` twice

arnaudstiegler replied:
Good catch! Removed it

ydshieh commented on src/transformers/models/trocr/modeling_trocr.py:
```diff
@@ -891,13 +891,40 @@ def forward(
+        >>> import requests
+        >>> from PIL import Image
 
+        >>> processor = TrOCRProcessor.from_pretrained("microsoft/trocr-base-handwritten")
```
This is better to be after showing the dummy model, i.e. after the line

```
>>> model = VisionEncoderDecoderModel(encoder=encoder, decoder=decoder)
```

arnaudstiegler replied:
Done

ydshieh commented on src/transformers/models/trocr/modeling_trocr.py:
```diff
@@ -891,13 +891,40 @@ def forward(
 
+        >>> processor = TrOCRProcessor.from_pretrained("microsoft/trocr-base-handwritten")
+
+        >>> # TrOCR is a decoder model and should be used within a VisionEncoderDecoderModel
```
Maybe add a comment here

```
>>> # init vision2text model with random weights
```

arnaudstiegler replied:
Done

ydshieh commented on src/transformers/models/trocr/modeling_trocr.py:
```diff
@@ -891,13 +891,40 @@ def forward(
+        >>> # load image from the IAM dataset
+        >>> url = "https://fki.tic.heia-fr.ch/static/img/a01-122-02.jpg"
+        >>> image = Image.open(requests.get(url, stream=True).raw).convert("RGB")
+        >>> pixel_values = processor(image, return_tensors="pt").pixel_values
```
Maybe put this line just before 

```
>>> text = "hello world"
```

arnaudstiegler replied:
Done

ydshieh commented on utils/documentation_tests.txt:
```diff
@@ -38,6 +38,7 @@ src/transformers/models/speech_encoder_decoder/modeling_speech_encoder_decoder.p
 src/transformers/models/speech_to_text/modeling_speech_to_text.py
 src/transformers/models/speech_to_text_2/modeling_speech_to_text_2.py
 src/transformers/models/swin/modeling_swin.py
+src/transformers/models/trocr/modeling_trocr.py
```
nice ❤️!

Are you able to run the doctest locally and get it pass?

arnaudstiegler replied:
Yep, runs fine locally! I ran the test again after addressing your comments

ydshieh reviewed (approved):
@arnaudstiegler

Very nice & clean! I left a few tiny comments, but this PR is ready to be merged!

arnaudstiegler committed:
address comments

arnaudstiegler committed:
fix doctest

arnaudstiegler committed:
apply make fixup

arnaudstiegler commented:
Addressed the comments and re-ran the test locally!
There's one thing though: `make fixup` messes up the formatting of the docstring, and removes the additional blank at the end of the docstring ([here](https://github.com/huggingface/transformers/blob/ae63ca7033aa1d6220a855498a5403d846ba51b3/src/transformers/models/trocr/modeling_trocr.py#L933-L934)). The issue is that without the additional line, doctest will fail when comparing outputs. Not sure why to be honest, but do you know how to prevent that `make fixup` from doing this? Otherwise, next time the file is touched, make mixup will introduce a bug in the docstring

ydshieh commented:
Those blank lines issues should be treated by the places regarding `utils/prepare_for_doc_test.py` as shown in this guide [doc](https://github.com/huggingface/transformers/tree/main/docs#for-python-files).

This means you don't need to add extra blank line in the file. Did you run `python utils/prepare_for_doc_test.py` as in the guide before and after running the doctest?

arnaudstiegler commented:
Got it! I did run the script and it seems like it's not being applied to the trocr (but I can see the added lines on other model files). I'll debug that today

arnaudstiegler committed:
remove additional blank

arnaudstiegler committed:
fix file to have correct split for prepare_for_doc_test

arnaudstiegler commented:
> OK, thank you. Don't hesitate to report it if you find this is some bug in `prepare_for_doc_test.py`.

Figured out the issue: there was some formatting issue in one of the docstring (see last commit) that prevented the script to correctly add a line. From `utils.prepare_for_doc_test.process_doc_file`
```
splits = code.split("```")
splits = [s if i % 2 == 0 else process_code_block(s, add_new_line=add_new_line) for i, s in enumerate(splits)]
```
splits were incorrect because of that, and the script wouldn't add an additional \n
Now test run locally, and make fixup has been run as well.
Should be good to go

arnaudstiegler commented:
@ydshieh Is there anything else to do? I don't have writing access here, so I can't merge this

ydshieh commented on src/transformers/models/trocr/modeling_trocr.py:
```diff
@@ -891,13 +891,46 @@ def forward(
+        ...     ViTModel,
+        ...     TrOCRConfig,
+        ...     ViTConfig,
+        ... )
```
just a tiny thing, but would be really great if this could be in some order,  like

```python
        ...     TrOCRConfig,
        ...     TrOCRProcessor,
        ...     TrOCRForCausalLM,
        ...     ViTConfig,
        ...     ViTModel,
        ...     VisionEncoderDecoderModel,
```

arnaudstiegler replied:
Done

ydshieh commented on src/transformers/models/trocr/modeling_trocr.py:
```diff
@@ -891,13 +891,46 @@ def forward(
-
         >>> model = VisionEncoderDecoderModel(encoder=encoder, decoder=decoder)
+
+        >>> # If you want to start from the pretrained model, load it from a VisionEncoderDecoderModel
```
Think would be better described as

```
>>> # If you want to start from the pretrained model, load the checkpoint with `VisionEncoderDecoderModel`
```

arnaudstiegler replied:
Done

ydshieh commented on src/transformers/models/trocr/modeling_trocr.py:
```diff
@@ -891,13 +891,46 @@ def forward(
+        >>> # training
+        >>> model.config.decoder_start_token_id = processor.tokenizer.cls_token_id
+        >>> model.config.pad_token_id = processor.tokenizer.pad_token_id
+        >>> model.config.vocab_size = model.config.decoder.vocab_size
```
```suggestion
        >>> model.config.vocab_size = model.config.decoder.vocab_size

```

ydshieh replied:
(just a new line for style reason)

ydshieh reviewed (commented):
Hi, @arnaudstiegler,

I have left a few more comments: most of them are just some nits.
This PR is really ready, and I will merge once you can apply the suggestions 💯 
Thank you so much!

ydshieh commented on src/transformers/models/trocr/modeling_trocr.py:
````diff
@@ -597,8 +597,8 @@ def forward(
                 that don't have their past key value states given to this model) of shape `(batch_size, 1)` instead of
-                all ``decoder_input_ids``` of shape `(batch_size, sequence_length)`. inputs_embeds (`torch.FloatTensor`
-                of shape `(batch_size, sequence_length, hidden_size)`, *optional*): Optionally, instead of passing
+                all `decoder_input_ids` of shape `(batch_size, sequence_length)`. inputs_embeds (`torch.FloatTensor` of
````
very nice!

arnaudstiegler committed:
Update src/transformers/models/trocr/modeling_trocr.py

Co-authored-by: Yih-Dar <2521628+ydshieh@users.noreply.github.com>

arnaudstiegler committed:
address comments

arnaudstiegler committed:
changing text | adding loss check | make fixup

arnaudstiegler committed:
make fixup

arnaudstiegler commented on src/transformers/models/trocr/modeling_trocr.py:
```diff
@@ -891,13 +891,46 @@ def forward(
+        >>> model.config.vocab_size = model.config.decoder.vocab_size
+        >>> labels = processor.tokenizer(text, return_tensors="pt").input_ids
+        >>> outputs = model(pixel_values, labels=labels)
+        >>> loss = outputs.loss
```
Added the loss. I actually changed the text to reflect the actual text on the image because the loss was super high with the dummy text (loss=22.00 vs loss=4.00). Let me know what's best

ydshieh replied:
Very nice for the change of target text! Awesome 👍

arnaudstiegler commented:
Rebased on latest main and addressed the comments, the failing test is coming from main (failing on main as well)

ydshieh commented:
Hi, the failed check seems irrelevant to your PR. No need to fix it here 😃. I will give a final look and merge. Thank you 💗

arnaudstiegler commented:
Alright, thank you!

ydshieh commented:
By the way, may I wonder why you choose to display 3 decimal for the loss value? O remember ee use 2 decimal in doc.py. I will check once I am available.

ydshieh commented on src/transformers/models/trocr/modeling_trocr.py:
```diff
@@ -891,13 +891,52 @@ def forward(
+        >>> labels = processor.tokenizer(text, return_tensors="pt").input_ids
+        >>> outputs = model(pixel_values, labels=labels)
+        >>> loss = outputs.loss
+        >>> round(loss.item(), 3)
+        5.303
```
```suggestion
        >>> round(loss.item(), 2)
        5.30
```

ydshieh commented on src/transformers/models/trocr/modeling_trocr.py:
```diff
@@ -891,13 +891,52 @@ def forward(
+        >>> import requests
+        >>> import torch
+        >>> from PIL import Image
+
+        >>> torch.manual_seed(0)  # doctest: +IGNORE_RESULT
+
```
```suggestion

```

ydshieh commented on src/transformers/models/trocr/modeling_trocr.py:
```diff
@@ -891,13 +891,52 @@ def forward(
+        ...     VisionEncoderDecoderModel,
+        ... )
+        >>> import requests
+        >>> import torch
```
```suggestion
```

ydshieh reviewed (commented):
Just added 3 final suggestions

- No (real) need to set the seed
- display only 2 decimal numbers

This is more about to be aligned with the code sample in `doc.py`.

arnaudstiegler committed:
Update src/transformers/models/trocr/modeling_trocr.py

Co-authored-by: Yih-Dar <2521628+ydshieh@users.noreply.github.com>

arnaudstiegler committed:
Update src/transformers/models/trocr/modeling_trocr.py

Co-authored-by: Yih-Dar <2521628+ydshieh@users.noreply.github.com>

arnaudstiegler committed:
Update src/transformers/models/trocr/modeling_trocr.py

Co-authored-by: Yih-Dar <2521628+ydshieh@users.noreply.github.com>

arnaudstiegler committed:
make fixup

arnaudstiegler commented:
Applied your changes, tested them locally, and ran make fixup as this was needed after the changes

arnaudstiegler commented:
Ok, I tried rebasing on latest master and it doesn't seem to be doing the trick. Not sure what's causing that unfortunately, but unlikely to be due to the code changes I've done :)

ydshieh commented:
Thank you again @arnaudstiegler ! (also for your patience)

I merge this PR now ❤️.
